.github/workflows/db_import.yml

```diff
@@ -0,0 +1,65 @@
+name: DB Import
+
+on:
+  pull_request:
+    branches:
+      - main
+    paths:
+      - devtools/db_import/**
+  # This also needs to run on push to fill the GitHub Actions Cache with the current Docker image,
+  # so that subsequent PRs can inherit this cache and leverage the prebuilt Docker image.
+  push:
+    branches:
+      - main
+
+jobs:
+  testing:
+    runs-on:
+      # These tests need to run on GCE because the integration tests
+      # access the GCS buckets which has to be authenticated with a
+      # service account.
+      - self-hosted  # must come first
+      - runner-group=${{ github.event_name == 'pull_request' && 'presubmit' || 'postsubmit' }}
+      - environment=prod
+      - cpu
+      - os-family=Linux
+    steps:
+      - name: "Checking out PR repository"
+        uses: actions/checkout@e2f20e631ae6d7dd3b768f56a5d2af784dd54791  # v2.5.0
+
+      - name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@4c0219f9ac95b02789c1075625400b2acbff50b1 # v2.9.1
+
+      # This builds the Docker image from the Dockerfile on each invocation of
+      # the workflow. This has the advantage that we don't have to deal with
+      # outdated Docker images in registries. Any changes to the Dockerfile
+      # or the context will be automatically reflected in the used image.
+      #
+      # This is of course time consuming, so we use the GitHub Actions Caching
+      # Service to make the build step fast (15 mins uncached build time,
+      # 15secs cached build time). If needed it's also possible to push
+      # newly built images to a registries, but this should only happen in
+      # postsubmit.
+      - name: Build Docker image
```

Comment:
Ah... This is super cool. Thanks for researching this method.

I'm slightly worried about two points:

1. For reproducibility, users might want to find the same `image@hash` to pull from a registry as CI, is it still possible (or easy) with Github action cache?
2. It looks like the max total cache size of a repo is 10GB (https://docs.github.com/en/actions/using-workflows/caching-dependencies-to-speed-up-workflows#usage-limits-and-eviction-policy), I'm wondering if we max out soon and see cache misses. (I can see ~700MB is used from this PR).

Regarding the Github cache, although it is safe to use now, I have a plan to change the security model of our CI (in order to improve efficiency) which might make it unsafe to use (cache poisoning from untrusted runners). I'm happy to explain it more but it's another reason we might not want to use it right now.

Just FYI (you might already know), in this repo we have added the IREE-style docker image managing tool: https://github.com/openxla/openxla-benchmark/blob/main/devtools/docker/manage_images.py. You add docker images to build in `devtools/docker/image_deps.json` and `manage_images.py` will build and update the references in repo.

Reply:
Regarding 1. Yes that's possible and easy to do, just not implemented yet. The usual approach would be to also push to a registry when running from main (no PR), so that users can get the container from the registry. I haven't implemented that yet, because it needs authentication.

Regarding 2. Yes, total cache size is a concern, especially when we use that technique for more than this one docker image. But still, as long as we don't make frequent changes to this docker image, the total cache usage shouldn't go beyond these 700MiB. Once this PR gets merged, this PR's cache gets evicted and the postsubmit workflow will create a cache on the main branch. This cache will then be reused for all following PRs. As long as there is no change to the Docker container, no new cache artifacts are created. The cache usage can be further reduced by not pushing to the cache from the main branch, but instead only push to the registry from the main branch. Each PR will then start with an empty cache and download the cached layers from the registry. If there were changes to the Docker image definition, the PR's cache would preserve the image between CI runs. That way the cache only needs to do its thing for the lifetime of a PR - and only for PRs that actually make changes to a cached Docker image.

Regarding cache poisoning: I'm not sure if I understand what your plans are, but I think cache poisining is not a big issue with GitHub caches, because the cache if not shared. Each PR and branch have their own cache, meaning one PR can't affect the cached artifacts of another PR. But when a new PR gets created its cache gets prefilled with the main branch's cache, so that you don't start with an empty cache.

Regarding `manage_images.py`. Yes, I saw that tool and I'm happy to switch to that if you prefer that. I wanted though to give `docker buildx` with its enhanced caching a try. It makes changes to the Docker container fully self-served. No need for a maintainer to push a new image or update a digest. Everything gets tested on the CI and it's possible to make changes to a Docker image and some other part of the repo at the same time in the same commit.

Reply:
I just realized I might be wrong on how cache keying actually works since Docker Buildx is directly talking to the caching API. Let me double check that.

Reply:
> The usual approach would be to also push to a registry when running from main (no PR), so that users can get the container from the registry. I haven't implemented that yet, because it needs authentication.

How do users find the digest to pull the image? Currently that is clear because we embed the digests in the workflow files and they are tracked in git history. If we don't track digests in commits (and just have the tag `presubmit:current` for the latest one), it might be a problem to reproduce the exact environment at a previous commit, as users don't know which digest was used in that commit.

> Yes, total cache size is a concern, especially when we use that technique for more than this one docker image. But still, as long as we don't make frequent changes to this docker image, the total cache usage shouldn't go beyond these 700MiB.

I prefer to have one method to manage all docker images in a repo, which means if we move to use `docker/build-push-action`, we might want to upload all docker images to the cache. This might quickly saturate the cache limit. For example the image we are using in benchmark `gcr.io/iree-oss/openxla-benchmark/openxla-benchmark-cuda11.8-cudnn8.9` takes 18.2GB (I guess due to CUDA)

For the security concern, right now our CI security model assumes that all postsubmit jobs are running on trusted runners. That creates some problems, for example, the testing/benchmark runners can't be permanent and shared between pre and postsubmit (for efficiency), we can't accept external partners to contribute testing/benchmark runners. Theoretically, we should only need to ensure the building and release jobs are running on trustable runners (so our artifacts are trustable), the testing and benchmark jobs can be running on untrustable ones (since they only upload JSON files, which don't pose much risk).

This change means we will mix trustable and untrustable jobs in postsubmit workflows. So if Github action cache can't prevent untrustable jobs in the same workflow to poison caches, it can be a blocker for us to move to this less-restricted security model. And AFAIK that's the case (https://scribesecurity.com/blog/github-cache-poisoning).

The security concern shouldn't be a deal breaker here since the new security model is still in design. But I'd like to raise it here for the awareness.

Reply:
> > The usual approach would be to also push to a registry when running from main (no PR), so that users can get the container from the registry. I haven't implemented that yet, because it needs authentication.
> 
> How do users find the digest to pull the image? Currently that is clear because we embed the digests in the workflow files and they are tracked in git history. If we don't track digests in commits (and just have the tag `presubmit:current` for the latest one), it might be a problem to reproduce the exact environment at a previous commit, as users don't know which digest was used in that commit.

Users would rebuild the docker container from source using `docker buildx build` (probably wrapped in a script). And `docker buildx` would determine hashes for each layer and then pull those layers from the registry (or actually build them from source if they are not cached.)

> 
> > Yes, total cache size is a concern, especially when we use that technique for more than this one docker image. But still, as long as we don't make frequent changes to this docker image, the total cache usage shouldn't go beyond these 700MiB.
> 
> I prefer to have one method to manage all docker images in a repo, which means if we move to use `docker/build-push-action`, we might want to upload all docker images to the cache. This might quickly saturate the cache limit. For example the image we are using in benchmark `gcr.io/iree-oss/openxla-benchmark/openxla-benchmark-cuda11.8-cudnn8.9` takes 18.2GB (I guess due to CUDA)

Using a registry for caching (https://docs.docker.com/build/cache/backends/registry/) instead of GitHub Actions could solve that. I'm not sure though how garbage collection works in that case.

What would it entail for me to switch to the digests tool? I guess you have to build the image and upload it and send me a digest which I can put into this PR, right?

> For the security concern, right now our CI security model assumes that all postsubmit jobs are running on trusted runners. That creates some problems, for example, the testing/benchmark runners can't be permanent and shared between pre and postsubmit (for efficiency), we can't accept external partners to contribute testing/benchmark runners. Theoretically, we should only need to ensure the building and release jobs are running on trustable runners (so our artifacts are trustable), the testing and benchmark jobs can be running on untrustable ones (since they only upload JSON files, which don't pose much risk).
> 
> This change means we will mix trustable and untrustable jobs in postsubmit workflows. So if Github action cache can't prevent untrustable jobs in the same workflow to poison caches, it can be a blocker for us to move to this less-restricted security model. And AFAIK that's the case (https://scribesecurity.com/blog/github-cache-poisoning).
> 
> The security concern shouldn't be a deal breaker here since the new security model is still in design. But I'd like to raise it here for the awareness.

Not sure if there is a way to restrict write access to the cache based on runner or workflow. If so that would solve it, I guess?. Another option is to use registry based caching and GCE machine authentication. In that case only our trusted runners would get write access to the registry/cache.

Reply:
Agree, I feel like most of the problems are related to Github action cache and `docker/build-push-action` is actually working well. Many problems can probably be solved by switching registry to our own gcr with GCE machine authentication.

Reply:
Considering our requirements I see 2 ways forward:

1. Rewrite it to use the proven method (`prod_digests.txt`)
2. Move to registry based caching and move the cuda container to use the same (probably in a separate PR)

Reply:
Moved to prod_digests.txt. Thank you for the discussion. :-)